### PR TITLE
move: manage [addresses] automatically on publish / upgrade

### PR DESCRIPTION
**Pull request by rvantonder**
## Description 

This PR enables automation of package addresses so that users don't need to fiddle with the `[addresses]` section when publishing/upgrading a package. However, if they are accustomed to the previous habits of changing `[addresses]`, they can still do so: the PR is backwards compatible with the "old way" described in the docs, which we'll keep during transition. In that case, though, they can't benefit from publishing to multiple chains.

There are two key implementation changes:

- **Part 1** (see inline comment): For named addresses in `[addresses]` we resolve the original published ID from the `Move.lock` if it exists there under some `chain-id`. 
  - Crucially, we assume that a key/value such as `foo = 0x0` with the special value of `0x0` implies that `foo` is the current package for which we can substitute modules for addresses to-be-published.

- **Part 2** (remainder of changes): When a package is already published, or is to be upgraded, there is an existing non-`0x0` original ID in the `Move.lock` associated with the package. To make progress, we have to substitute the `0x0` address at compile time for module names of the updated package that is to be published or upgraded. After compiling, we restore the original ID to its prior value that it was published at. This essentially automates what a user would have had to do manually. This automated part uses the `Move.lock` file as the working space for respective `chain-id`s. The top-level logic happens for `Publish` and `Upgrade` client commands.

## Test plan 

Existing tests + updated upgrade happy path test to exercise Part 1 and Part 2 above + local testing against localnet, testnet

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-package/src/resolution/resolution_graph.rs [hunk 1]
```diff
@@ -345,19 +349,52 @@ impl Package {
...
+            if *addr == Some(AccountAddress::ZERO) {
+                // The address in the manifest is set to 0x0, meaning `name` is associated with 'this'
+                // package. Published dependent package IDs are resolved by `chain_id` from the
+                // `Move.lock` when a package is to be published or upgraded.
+                if let Some(original_id) = self.resolve_original_id_from_lock(chain_id) {
+                    let addr = AccountAddress::from_str(&original_id)?;
+                    resolving_table.define((pkg_id, *name), Some(addr))?;
+                    continue;
+                }
+            }
```
Comment (rvantonder) [inline external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:380]: This is Part 1: identify when we should resolve original IDs from the `Move.lock`, and do so when possible for the respective `chain_id`.

Nitty gritty note: the first commit in this PR propagates `chain_id` to this part of the logic: ff73dbfbc8e759c8c3ffc44a7299280f34d6d12a. It might work to pass this value in `BuildConfig` instead of propagating through function args. The change for propagating through functions didn't feel too invasive though, so went with that.

File: crates/sui/src/client_commands.rs [hunk 2]
```diff
@@ -946,20 +965,39 @@ impl SuiClientCommands {
...
+                // Restore original ID, then check result.
+                if let Ok(Some(previous_id)) = previous_id {
+                    let _ = sui_package_management::set_package_id(
+                        &package_path,
+                        build_config.clone().install_dir,
+                        chain_id,
+                        previous_id,
+                    );
+                }
+                let (dependencies, compiled_modules, _, _) = compile_result?;
```
Comment (rvantonder) [inline crates/sui/src/client_commands.rs:1000]: I don't _love_ the how setting/resetting the original-id here bulks up this code block, but it is straightforward to reason about so I didn't wrestle further with it. 

I couldn't find a way to avoid `clone` in some places and tried some things 🤷

File: external-crates/move/crates/move-package/src/lock_file/schema.rs [hunk 3]
```diff
@@ -346,6 +346,31 @@ pub enum ManagedAddressUpdate {
     },
 }
 
+/// Sets the `original-published-id` to a given `id` in the lock file. This is a raw utility
+/// for preparing package publishing and package upgrades. Invariant: callers maintain a valid
+/// hex `id`.
+pub fn set_original_id(file: &mut LockFile, environment: &str, id: &str) -> Result<()> {
```
Comment (rvantonder) [inline external-crates/move/crates/move-package/src/lock_file/schema.rs:352]: We discussed an in-memory representation of lock contents, and this kind of thing makes me eager to make that work instead of using the file. An in-memory representation needs to be a separate refactor though, and I think potentially really error-prone based on how things happen right now. So I went with the predictable, mechanical way of working via the lock file for now.

File: crates/sui/src/client_commands.rs (hunk 2 above)
Comment (amnn) [reply crates/sui/src/client_commands.rs:1000]: Come to think of it, if you are doing a publish, do you need to set the previous ID? Presumably it will get overwritten based on transaction effects below?

**Review by amnn — COMMENTED** [review]
Change looks mostly good, thanks @rvantonder -- main thing I want to poke at a bit more is the need to change the `BuildConfig` in `cli_tests.rs`, which hints at a potential actual issue.

File: crates/sui/src/client_commands.rs (hunk 2 above)
Comment (rvantonder) [reply crates/sui/src/client_commands.rs:1000]: Yes, need to reset because publishing can fail (insufficient gas, etc.) and will leave the state as `0x0` in that case. It gets overwritten only if publishing succeeds.

File: crates/sui-package-management/src/lib.rs [hunk 4]
```diff
@@ -94,6 +95,43 @@ pub async fn update_lock_file(
...
+/// The function returns the existing `original-published-id`, if any.
+pub fn set_package_id(
+    package_path: &Path,
+    install_dir: Option<PathBuf>,
+    chain_id: Option<String>,
+    id: AccountAddress,
+) -> Result<Option<AccountAddress>, anyhow::Error> {
+    let lock_file_path = package_path.join(SourcePackageLayout::Lock.path());
+    if let Ok(mut lock_file) = File::open(lock_file_path.clone()) {
+        if let Some(chain_id) = chain_id {
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply crates/sui-package-management/src/lib.rs:112]: We could and I thought about it--in this case, it means more work for the caller (`if Some(chain_id)...`, which I'm fine with but in this case it means the `client_commands` code block grows more bulky. I'll go ahead with the suggestion though

File: crates/sui/src/client_commands.rs [hunk 5]
```diff
@@ -861,28 +861,48 @@ impl SuiClientCommands {
...
+                    build_config.clone(),
+                    &package_path,
+                    upgrade_capability,
+                    with_unpublished_dependencies,
+                    skip_dependency_verification,
+                    env_alias,
+                )
+                .await;
+                // Restore original ID, then check result.
+                if let Ok(Some(previous_id)) = previous_id {
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply crates/sui/src/client_commands.rs:895]: Fair! I was debating if/how to handle hard errors in `set_package_id`--will return early in the error case (basically, when we can't update the lock file)

File: crates/sui/tests/cli_tests.rs [hunk 6]
```diff
@@ -1876,7 +1876,10 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
 
     // Create a new build config for the upgrade. Initialize its lock file
     // to the package we published.
-    let build_config_upgrade = BuildConfig::new_for_testing().config;
+    let mut build_config_upgrade = BuildConfig::new_for_testing().config;
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply crates/sui/tests/cli_tests.rs:1879]: Ah, missed walking through this, but also tried handling this in a more principled way, considering your comment--tl;dr with long incoming thoughts:

TL;DR I think it's more a matter of `install_dir` getting in the way for publish/upgrade flows where we would never reasonable expect (or allow the user to put) a `Move.lock` other than the root directory for publishing/upgrading. Although we can (and I tried) defining a behavior around `install_dir` in these cases to be more hygienic for tests, it didn't pan out and needs more effort, where the benefit is dubious.

---

Thoughts:

- When we publish/upgrade, I don't believe there is a notion of a user-configurable `install_dir`, that's an option for `sui move build`. Which is to say, we currently expect the `Move.lock` to be in the root package dir, and `build/lock` workspace relative to the package root, when we publish/upgrade.

- As pointed out, the `install_dir` isn't being "respected" (if we decided to define respecting as meaning "we should access the lock file under `install_dir` always")
  - Though again, in this test, `install_dir` is set in a way that isn't possible via the CLI or for publishing AFAIK. If there were I might be more concerned about changing the inner values here, but with the current state the idea of `install_dir` being anything other than the the current directory on publish/upgrade is a kind of undefined behavior.

  - So why can't we just use this `install_dir` from `BuildConfig::new_from_testing` as-is and what would it take to "respect" it in publish/upgrade flows? 
    - I dug into this for quite some hours and felt like I got close to succeeding (didn't quite) and felt this is a big hassle for something that is only going to benefit a test setup scenario and not any realistic publish/upgrade flows. One reason is that when publishing and resolving IDs, [we use the package path](https://sourcegraph.com/github.com/MystenLabs/sui@cccd6409d90c9460d80892e83e6d18d72dc27f62/-/blob/crates/sui-package-management/src/lib.rs?L115), and would now need to use the custom-set `lock_path` in build config instead (and painfully propagate that through). Similar changes in this PR and for test setup.
      - Made lots of related changes and things still didn't work out (e.g., publish/upgrade actions still looking in the package dir for `Move.lock`), started questioning my sanity and if this is really worth it.

So, I think it's reasonable but unnecessary to use these values in `BuildConfig` for test purposes, but I'm fairly convinced that the operation of setting IDs in the `Move.lock` during publish/upgrade is disjoint from build configs/`install_dir` concerns, and okay to punt on needing that to work in this test.

**Comment by rvantonder** [comment]
> I want to poke at a bit more is the need to change the BuildConfig in cli_tests.rs, which hints at a potential actual issue.

It was a good thing to poke at! I reasoned through this a bit more deeply. I think we can be cleaner about some of the state management but that it's not an actual issue--rerequesting review given thoughts in https://github.com/MystenLabs/sui/pull/18470#discussion_r1663380854

File: crates/sui/tests/cli_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui/tests/cli_tests.rs:1879]: > When we publish/upgrade, I don't believe there is a notion of a user-configurable `install_dir`

I don't think this is true -- it is configurable using the `--install-dir` flag on `sui client publish` and `upgrade`. It's part of the `build_config: MoveBuildConfig` parameter that is flattened into the `Publish` and `Upgrade` commands.

[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui/tests/cli_tests.rs:1879]: I believe the main reason why `install_dir` exists in `BuildConfig` is precisely for testing purposes -- so that when we write tests that require building Move packages in the repo, it doesn't dirty the repo state.

In CI this is especially important in the case of lock files because lock files rely on atomic rename to be durable, and that property only holds if the rename is within the same mount point. On CI, the temp directory uses tmpfs, so that's not true, and that is a major reason why `BuildConfig::new_for_testing()` currently exists.

The detail about us reading the lock file relative to the package path is an interesting one. My argument is that this does seem like the right thing to do for a dependency, because the only valid location for the lock file of a dependency is right next to its `Move.toml` (and in general we don't have a way to specify an install directory for dependencies, only the root package being built), but if we are using this same technique to look for the lock file of the root package, that's when we have a problem.

For me that means that if we're in this situation because our test requires setting up a dependency with its lock file and then using that in a future build, then the right thing to do is to construct our own `BuildConfig` without using a temp directory for the install dir (when building the dependency), as long as we are copying the dependency's package contents into a temp directory before doing so.

However, if we're in this situation because we are trying to read the root package's lock file from the wrong place then we should re-consider the implementation or test set-up.

Ultimately the thing that is strange about the code is to use the `new_for_testing` constructor but then undo the main thing that makes it different from the default `BuildConfig`.

File: crates/sui/tests/cli_tests.rs [hunk 7]
```diff
@@ -1940,7 +1943,7 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
 }
 
 #[sim_test]
-async fn test_package_management_on_upgrade_command_conflict() -> Result<(), anyhow::Error> {
+async fn xtest_package_management_on_upgrade_command_conflict() -> Result<(), anyhow::Error> {
```
Comment (amnn) [inline crates/sui/tests/cli_tests.rs:1946]: spurious?

File: crates/sui/tests/cli_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui/tests/cli_tests.rs:1879]: I see now that the dependency in this test is slightly different still -- it's to do with the lock file carrying version information between a publish and an upgrade of a package (in hindsight, it makes perfect sense that that's what we're talking about here).

To me this leans slightly more in the first camp -- that if we have been supplied an `install_dir` for a package, then that is where the upgrade command we should look for its lock file. I.e. the ideal way to write this test is to:

- Create a temp directory to act as the install dir.
- Build and publish the first version of the package, without copying it from the repo, but using the temp install dir.
- Build the second version of the package and upgrade it, again without copying it from the repo, but using the same build config, with the same install dir.

**Review by amnn — COMMENTED** [review]
Sorry for the rambling thoughts about this whole BuildConfig thing -- PTAL at the suggestion, I'm hopeful there's a simple way around this, because unfortunately this flag is user-configurable.

File: crates/sui/tests/cli_tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rvantonder) [reply crates/sui/tests/cli_tests.rs]: > Sorry for the rambling thoughts about this whole BuildConfig thing

That's totally fine, useful context and thoughts, trying to figure out what's best here.

> I don't think this is true -- it is configurable using the `--install-dir`

TIL!! I was wrong, I stand corrected. Okay. So, after going through the thoughts and revisiting test setup, I still think the "this behavior is not user-configurable and not worth chasing a fix the mainline implementation". Because: AFAIK, there is actually no reason to care or honor installing `Move.lock` to `install_dir` (even if `install_dir` is user-configurable). AFAIK we only `lock.commit(somewhere)` where `somewhere` is what [the `lock_file` path is specified](https://sourcegraph.com/github.com/MystenLabs/sui@94a8236a61dbe90f1f9f8eea9364b12868670ff0/-/blob/external-crates/move/crates/move-package/src/lib.rs?L71-73). In the mainline execution we don't set `lock_file` relative to `install_dir` (we _do_ create `build/locks` relative to `install_dir`--the working space before committing locks). `lock_file` value is not a user-configurable value (this time I checked and we see clap skip).

All to say, the decision in the test setup to set the [`lock_file` path to `install_dir.join("Move.lock")`](https://sourcegraph.com/github.com/MystenLabs/sui@94a8236a61dbe90f1f9f8eea9364b12868670ff0/-/blob/crates/sui-move-build/src/lib.rs?L88) seems arbitrary wrt to `install_dir`, it could be any tmp dir. But presumably because that test setup conveniently creates a tmp dir for `install_dir`, it just sets the `lock_path` there as well. Point is, if a user configured `install-dir`, there isn't, AFAIK, a real need or intention to actually install `Move.lock` there for them--that's something this test setup uniquely decides to do, and presumably because the `Move.lock` can be checked for testing if it's output there.


Re suggestion:

> that if we have been supplied an install_dir for a package, then that is where the upgrade command we should look for its lock file.

So because of ^ I don't think it's the case that if we have an `install_dir` for a package, then we should look for the lock file there. In fact, when I realized that, it turns out I can remove the line that changes `install_dir` and keep the one in build config:

```patch
- build_config_upgrade.install_dir = Some(install_dir.clone());
```

And the test still passes. It's the case if we have a `lock_file` path (which is not user-configurable), that we should look for the lock file there. I.e., this line is important:

```patch
+ build_config_upgrade.lock_file = Some(pkg_root.join("Move.lock"));
```

To wit, and re: your suggesting--our mainline logic [wants to find the lock file in the source package root](https://sourcegraph.com/github.com/MystenLabs/sui@cccd6409d90c9460d80892e83e6d18d72dc27f62/-/blob/crates/sui-package-management/src/lib.rs?L115) not at the `lock_file` path (which may be overridden in some internal way, like in the test, but cannot realistically be overridden by a user otherwise). Changing the mainline code to always use `lock_file` path instead of "expect `Move.lock` in root of package" is the part I attempted to fix in the mainline code, turned out invasive, overly difficult, and eventually felt unnecessary for realistic use cases. That's why this test has the (perhaps not ideal) change to essentially say "no, just point `lock_path` to `Move.lock` at the package root").

I've updated the implementation in a way I think works a647c7f3fbcb8378d63250b4cc53c732a2f0e5ae. It is similar but simpler than before:

- Just one build config
- Keep `install_dir` in build config the way it is (it remains the working space for tmp lock files)
- Allow `publish` to put it's resulting `Move.lock` at the `BuildConfig` `lock_path` (i.e., `install_dir/Move.lock`)
- Set up the package for upgrading in a tmp directory (no clobbering) and copy the `install_dir/Move.lock` to the package root for upgrading.
- Point the build config `lock_file` to the `root/Move.lock` so it can be used as the input / output.
- Comment asserting that we simply don't support arbitrary `lock_file` in BuildConfig when that `lock_file` path needs to be used as an input (i.e., for upgrades). As explained above, supporting this `lock_file` to be arbitrary for the purposes of testing is really difficult (I tried). If we insist on asserting something more general re: arbitrary `lock_file` I can look at it in a follow up, but I won't lose sleep over assuming the `Move.lock` needs to be in a root directory for upgrades, and with these tests, we do that action in a tmp dir for the package to not clobber things.

Given my assumptions on "`Move.lock` in root" I think it would be fair to avoid the `BuildConfig::new_for_testing` and just use a sandbox for publish/upgrade where everything assumes the package root, but, like, the test achieves what it should at this point so `¯\_(ツ)_/¯`

[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui/tests/cli_tests.rs]: Thanks for digging in here! Main thing I'm taking away from this is that this space may need a general refactor, because this is too tough to navigate... 

> Given my assumptions on "`Move.lock` in root" I think it would be fair to avoid the `BuildConfig::new_for_testing` and just use a sandbox for publish/upgrade where everything assumes the package root, but, like, the test achieves what it should at this point so ¯\_(ツ)_/¯

Agreed -- simulating a "real" in-place upgrade where everything lives in a temp directory is probably the cleanest way to go (`new_for_testing` is really designed for the case where we're building a Move Package but don't care at all about the lock file or other artifacts -- just the in-memory representation of the package). But let's treat that as a follow-up.

**Review by amnn — APPROVED** [review]
Good to go, thanks for sticking with me!
